File: AirHockeyLightClient/AirHockeyLightClient/Noyau/Arbre/Visiteurs/VisiteurScale.swift
///////////////////////////////////////////////////////////////////////////////
/// @file VisiteurScale.swift
/// @author Pierre To
/// @date 2017-10-22
/// @version 0.1
///
/// @addtogroup log3900 LOG3900
/// @{
///////////////////////////////////////////////////////////////////////////////

import SceneKit
import GLKit

///////////////////////////////////////////////////////////////////////////
/// @class VisiteurScale
/// @brief Permet de scaler les noeuds
///
///
/// @author Pierre To
/// @date 2017-10-22
///////////////////////////////////////////////////////////////////////////
class VisiteurScale: VisiteurAbstrait {
    
    /// Vecteur trois dimensions pour le redimensionnement
    private var scale: GLKVector3
    
    /// Bool d'ajustement pour le redimensionnement
    private var ajusterScaling: Bool
    
    /// Constructeur
    init(scale: GLKVector3, ajusterScaling: Bool) {
        self.scale = scale
        self.ajusterScaling = ajusterScaling
    }
    
    /// Cette fonction permet d'assigner les facteurs de redimensionnement
    func assignerScale(scale: GLKVector3) {
        self.scale = scale
    }

    /// Visiter un accélérateur pour le rendre sélectionnable
    func visiterAccelerateur(noeud: NoeudAccelerateur) {
        if (noeud.estSelectionne()) {
            var scale = GLKVector3Add(
                            SCNVector3ToGLKVector3(noeud.scale),
                            (self.ajusterScaling ? GLKVector3DivideScalar(self.scale, 6.0) : self.scale))
            
            if (scale.x <= 1.0) {
                scale = GLKVector3.init(v: (1.0, 1.0, 1.0))
            }
            
            noeud.scale = SCNVector3FromGLKVector3(scale)
            
            // Envoyer la commande
            FacadeModele.instance.obtenirEtatEdition().currentUserObjectTransformChanged(uuid: noeud.obtenirUUID(),
                                                                                         pos: noeud.position, rotation: noeud.rotation.w, scale: noeud.scale)
        }
    }
    
    /// Visiter un maillet pour le rendre sélectionnable
    //virtual void visiterMaillet(NoeudMaillet* noeud);
    
    /// Visiter une table pour le rendre sélectionnable
    func visiterTable(noeud: NoeudTable) {
        // Ne rien faire
    }
    
    /// Visiter un point de contrôle pour le rendre sélectionnable
    func visiterPointControl(noeud: NoeudPointControl) {
        // Ne rien faire
    }
    
    /// Visiter un mur pour le rendre sélectionnable
    func visiterMur(noeud: NoeudMur) {
        if noeud.estSelectionne() {
            var scale = GLKVector3Add(SCNVector3ToGLKVector3(noeud.scale), self.scale)
            
            if (scale.z <= 1.0) {
                scale.z = 1.0
            }
            
            noeud.scale = SCNVector3.init(x: 1.0, y: 1.0, z: scale.z)
            
            // Envoyer la commande
            FacadeModele.instance.obtenirEtatEdition().currentUserObjectTransformChanged(uuid: noeud.obtenirUUID(),
                                                                                         pos: noeud.position, rotation: noeud.rotation.w, scale: noeud.scale)
        }
    }
    
    /// Visiter un portail pour le rendre sélectionnable
    func visiterPortail(noeud: NoeudPortail) {
        if noeud.estSelectionne() {
            var scale = GLKVector3Add(
                SCNVector3ToGLKVector3(noeud.scale),
                (self.ajusterScaling ? GLKVector3DivideScalar(self.scale, 12.0) : self.scale))
            
            if (scale.x <= 1) {
                scale = GLKVector3.init(v: (1.0, 1.0, 1.0))
            }
            
            noeud.scale = SCNVector3FromGLKVector3(scale)
            
            // Envoyer la commande
            FacadeModele.instance.obtenirEtatEdition().currentUserObjectTransformChanged(uuid: noeud.obtenirUUID(),
                                                                                         pos: noeud.position, rotation: noeud.rotation.w, scale: noeud.scale)
        }
    }
    
    //virtual void visiterRondelle(NoeudRondelle* noeud);
    
}

///////////////////////////////////////////////////////////////////////////////
/// @}
///////////////////////////////////////////////////////////////////////////////

File: Sources/DLL/Arbre/Visiteurs/VisiteurScale.cpp
///////////////////////////////////////////////////////////////////////////////
/// @file VisiteurScale.cpp
/// @author Nam Lesage
/// @date 2016-09-28
/// @version 0.1
///
/// @addtogroup inf2990 INF2990
/// @{
///////////////////////////////////////////////////////////////////////////////

#include "VisiteurScale.h"
#include "ModeleEtatJeu.h"

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurScale::VisiteurScale(const glm::vec3& scale, bool ajusterScaling)
///
/// Constructeur pour VisiterScalela
///
///	@param[in] scale : Vecteur des facteurs pour mise à l'échelle
///	@param[in] ajusterScaling : Si une mise à l'échelle est requise ou non
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
VisiteurScale::VisiteurScale(const glm::vec3& scale, bool ajusterScaling) 
	: scale_(scale), ajusterScaling_(ajusterScaling)
{
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurScale::~VisiteurScale()
///
/// Destructeur pour VisiterScale
///
///	@param[in] Aucun
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
VisiteurScale::~VisiteurScale() {

}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurScale::visiterAccelerateur(NoeudAccelerateur* noeud)
///
/// Cette fonction permet de visiter un accélérateur pour la mise en échelle.
///
///	@param[in] noeud : Un accélérateur
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurScale::visiterAccelerateur(NoeudAccelerateur* noeud) {
	if (noeud->estSelectionne())
	{
		glm::vec3 position = noeud->obtenirPositionRelative();
		glm::vec3 scale = noeud->obtenirScale() + (ajusterScaling_ ? (scale_ / 6.0f) : scale_);

		if (scale.x <= 1)
			scale = glm::vec3(1, 1, 1);

		noeud->scale(scale);
		sendTransform(noeud);

	}
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurScale::visiterMaillet(NoeudMaillet* noeud)
///
/// Cette fonction permet de visiter un maillet pour la mise en échelle.
///
///	@param[in] noeud : Un maillet
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurScale::visiterMaillet(NoeudMaillet* noeud) {
	// NE RIEN FAIRE
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurScale::visiterTable(NoeudTable* noeud)
///
/// Cette fonction permet de visiter une table pour la mise en échelle.
///
///	@param[in] noeud : Une table
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurScale::visiterTable(NoeudTable* noeud) {
	// NE RIEN FAIRE
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurScale::visiterPointControl(NoeudPointControl * noeud)
///
/// Cette fonction permet de visiter un point de contrôle pour la mise en échelle.
///
///	@param[in] noeud : Un point de contrôle
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurScale::visiterPointControl(NoeudPointControl * noeud) {
	// NE RIEN FAIRE
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurScale::visiterMur(NoeudMur* noeud)
///
/// Cette fonction permet de visiter un mur pour la mise en échelle.
///
///	@param[in] noeud : Un mur
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurScale::visiterMur(NoeudMur* noeud) {
	if (noeud->estSelectionne())
	{
		glm::vec3 position = noeud->obtenirPositionRelative();
		glm::vec3 scale = noeud->obtenirScale() + scale_;
		if (scale.z <= 1)
			scale.z = 1;

		noeud->scale(glm::vec3(1,1,scale.z));
		sendTransform(noeud);

	}
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurScale::visiterPortail(NoeudPortail* noeud)
///
/// Cette fonction permet de visiter un portail pour la mise en échelle.
///
///	@param[in] noeud : Un mur
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurScale::visiterPortail(NoeudPortail* noeud) {
	if (noeud->estSelectionne())
	{
		glm::vec3 position = noeud->obtenirPositionRelative();
		glm::vec3 scale = noeud->obtenirScale() + (ajusterScaling_ ? (scale_ / 12.0f) : scale_);

		if (scale.x <= 1)
			scale = glm::vec3(1, 1, 1);

		noeud->scale(scale);
		sendTransform(noeud);
	}
}

void VisiteurScale::visiterRondelle(NoeudRondelle * noeud)
{
}

void VisiteurScale::sendTransform(NoeudAbstrait* node)
{
	TransformEventCallback callback = ModeleEtatJeu::obtenirInstance()->getTransformEventCallback();
	if (callback)
	{
		callback(node->getUUID(), glm::value_ptr(node->obtenirPositionRelative()),node->obtenirRotation().y, glm::value_ptr(node->obtenirScale()));
	}
}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurScale::assignerScale(const glm::vec3 & scale)
///
/// Cette fonction permet d'assigner les facteurs de redimensionnement.
///
///	@param[in] scale : Vecteur des facteurs de redimensionnement
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurScale::assignerScale(const glm::vec3 & scale)
{
	scale_ = scale;
}

///////////////////////////////////////////////////////////////////////////////
/// @}
///////////////////////////////////////////////////////////////////////////////